mirroring/handler.go

```diff
@@ -0,0 +1,230 @@
+	h.newLeavesBuffer = b
+}
+
+func (h *LogHandler) GetLocalPublicKey() string {
+	return h.metadata.PublicKey
+}
+
+func (h *LogHandler) GetLocalTreeSize() int64 {
+	return h.metadata.SavedMaxIndex + 1
+}
+
+func (h *LogHandler) GetRemoteTreeSize() (int64, error) {
+	a, err := GetLogInfo()
+	if err != nil {
+		return 0, err
+	}
+	return *a.TreeSize, nil
+}
+
+func (h *LogHandler) GetRemoteRootSignature() (strfmt.Base64, error) {
+	a, err := GetLogInfo()
+	if err != nil {
+		return nil, err
+	}
+	return *a.SignedTreeHead.Signature, nil
+}
+
+func (h *LogHandler) GetRemoteRootPublicKey() (string, error) {
+	keyResp, err := h.client.Tlog.GetPublicKey(nil)
+	if err != nil {
+		return "", err
+	}
+	publicKey := keyResp.Payload
+
+	return publicKey, nil
+}
+
+func (h *LogHandler) GetRootSignature() ([]byte, error) {
+	s, err := base64.StdEncoding.DecodeString(h.metadata.LogInfo.SignedTreeHead.Signature.String())
+	if err != nil {
+		return nil, err
+	}
+	return s, nil
+}
+
+func (h *LogHandler) GetLocalRootHash() string {
+	return *h.metadata.LogInfo.RootHash
+}
+
+func (h *LogHandler) GetRemoteRootHash() ([]byte, error) {
+	info, err := GetLogInfo()
+	if err != nil {
+		return nil, err
+	}
+	logRootBytes, err := base64.StdEncoding.DecodeString(info.SignedTreeHead.LogRoot.String())
+	if err != nil {
+		return nil, err
+	}
+	logRoot := trilliantypes.LogRootV1{}
+	err = logRoot.UnmarshalBinary(logRootBytes)
+	if err != nil {
+		return nil, err
+	}
+	return logRoot.RootHash, nil
+}
+
+func (h *LogHandler) SetRootHash(rootHash string) {
+	*h.metadata.LogInfo.RootHash = rootHash
+}
+
+func (h *LogHandler) SetPublicKey(publicKey string) {
+	h.metadata.PublicKey = publicKey
+}
+
+func (h *LogHandler) SetRootSignature(sig strfmt.Base64) {
+	*h.metadata.LogInfo.SignedTreeHead.Signature = sig
+}
+
+func (h *LogHandler) SetLocalTreeSize(treeSize int64) {
+	h.metadata.SavedMaxIndex = treeSize - 1
+}
+
+// if kind=="", get all kinds of leaves
+func (h *LogHandler) GetAllLeavesForKind(kind string) error {
+	str := viper.GetString("tree_file_dir")
+	if h.newLeavesBuffer != nil && len(h.newLeavesBuffer) != 0 {
+		return errors.New("leaf buffer is not empty, please sync by saving")
+	}
+	file, err := os.Open(str)
+	artifacts := make([]Artifact, 0)
+	if err != nil {
+		return err
+	}
+	defer file.Close()
+	leaf := Artifact{}
+	reader := bufio.NewReader(file)
+	var line string
+	for kind != "" {
+		line, err = reader.ReadString('\n')
+		if err != nil {
+			if err == io.EOF {
+				break
+			} else {
+				return err
+			}
+		}
+
+		err = json.Unmarshal([]byte(line), &leaf)
+		if err != nil {
+			return err
+		}
+		if leaf.Kind == kind {
+			artifacts = append(artifacts, leaf)
+		}
+	}
+	for kind == "" {
+		line, err = reader.ReadString('\n')
+		if err != nil {
+			if err == io.EOF {
+				break
+			} else {
+				return err
+			}
+		}
+
+		err = json.Unmarshal([]byte(line), &leaf)
+		if err != nil {
+			return err
+		}
+		artifacts = append(artifacts, leaf)
+	}
```

Comment:
It's not immediately obvious why we need to split these into two for loops instead of using the kind statement from within (as a filter):

`if leaf.Kind == kind || kind == ""`

or somesuch, what do you think?

Reply:
I split them for the reason that if kind == "", then separating them would allow the for loop not have to go through the condition check every time before an artifact is appended, making the code run a bit faster.

On second thought however, compiler optimization and branch prediction probably makes this split pointless, so I combined them in the next commit.